File: lib/features/object_detection/domain/object_detector.dart
import 'dart:typed_data';

import '../../../core/errors/result.dart';
import 'detection_result.dart';

/// Абстрактный интерфейс детектора объектов (§7.5 ТЗ).
abstract class ObjectDetector {
  /// Детекция объектов на изображении.
  ///
  /// [imageBytes] — байты изображения (будут ресайзнуты до 300×300 внутри).
  ///
  /// Возвращает список обнаруженных объектов после NMS (IoU=0.45),
  /// отфильтрованных по confidence ≥ 0.45 (или 0.35 для объектов безопасности).
  Future<Result<List<DetectionResult>>> detect(Uint8List imageBytes);

  /// Освобождает TFLite Interpreter и связанные ресурсы.
  Future<void> dispose();
}

File: src/features/objectDetection/domain/objectDetector.ts
/**
 * Абстрактный интерфейс детектора объектов.
 */

import { Result } from '@core/errors/result';
import { DetectionResult } from './detectionResult';

export interface ObjectDetector {
  /**
   * Запустить детекцию на изображении.
   * @param imageInput — путь к файлу или готовый RGB-тензор
   * @returns список обнаруженных объектов (уже после NMS и фильтрации)
   */
  detect(imageInput: string | Uint8Array): Promise<Result<DetectionResult[]>>;

  /** Загрузить модель в память (lazy init) */
  initialize(): Promise<Result<void>>;

  /** Освободить ресурсы модели */
  dispose(): Promise<void>;

  /** Загружена ли модель */
  isReady(): boolean;
}
